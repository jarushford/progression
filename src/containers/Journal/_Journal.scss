.journal-form {
  @include form();

  h2 {
    @include form-title();
  }

  .journal-entry {
    @include form-caption();
    height: 100px;
  }

  button {
    @include button();
    margin: 20px auto;
  }
}

.journal {
  border-bottom: 1px solid $ruleLine;
  margin: 0 10px;

  div {
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 10px;
    position: relative;

    h3 {
      @include pageTitle();
      font-size: 1.6rem;
      margin: 0px;
    }

    .fa-times {
      cursor: pointer;
      position: absolute;
      font-size: 1.5rem;
      color: $text;
      transition: .4s ease;
      top: 10px;
      right: 0px;

      &:hover {
        font-size: 1.6rem;
        color: tomato;
      }
    }
  }

  .milestone-cap {
    font-size: 1.2rem;
    color: $accent;
    margin: 0 0 20px 15px;
  }
}