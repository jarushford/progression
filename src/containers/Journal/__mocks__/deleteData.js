export const deleteDataThunk = jest.fn()