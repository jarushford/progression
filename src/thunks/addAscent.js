import { fetchAscentsThunk } from './fetchAscents'
import { clearAscents, setError } from '../actions'

export const addAscentThunk = (ascent) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/ascents/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(ascent),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add ascent')
      }
      await dispatch(clearAscents())
      dispatch(fetchAscentsThunk(ascent.user_id))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}