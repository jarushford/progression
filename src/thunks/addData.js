import { clearThunkHelper } from '../utils/thunkHelpers'
import { fetchDataThunk } from './fetchData'
import { setError } from '../actions'

export const addDataThunk = (data, type) => {
  return async (dispatch) => {
    try {
      const url = `http://localhost:3000/api/progressionusers/${type}s/new`
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(data),
        headers: {
          "Content-Type": "application/json"
        }
      })

      if (!response.ok) {
        throw Error(`Could not add ${type}`)
      }

      await dispatch(clearThunkHelper(type)())
      type === 'journal' || type === 'milestone'
        ? dispatch(fetchDataThunk(data.user_id, data.project_id, type))
        : dispatch(fetchDataThunk(data.user_id, '', type))

      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}