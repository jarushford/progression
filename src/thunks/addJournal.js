import { fetchJournalThunk } from './fetchJournal'
import { clearJournal, setError } from '../actions'

export const addJournalThunk = (entry) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/progressionusers/journal/new'
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify(entry),
        headers: {
          "Content-Type": "application/json"
        }
      })
      if (!response.ok) {
        throw Error('Could not add entry')
      }
      await dispatch(clearJournal())
      dispatch(fetchJournalThunk(entry.user_id, entry.project_id))
      return true
    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}