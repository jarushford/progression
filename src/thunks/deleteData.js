import { clearThunkHelper } from '../utils/thunkHelpers'
import { fetchDataThunk } from './fetchData'
import { setError } from '../actions'

export const deleteDataThunk = (item_id, user_id, project_id, type) => {
  return async (dispatch) => {
    try {
      const data = {item_id, user_id, project_id}
      let url
      type === 'journal' || type === 'milestone'
        ? url = `http://localhost:3000/api/progressionusers/${user_id}/projects/${project_id}/${type}s/${item_id}`
        : url = `http://localhost:3000/api/progressionusers/${user_id}/${type}s/${item_id}`

      const response = await fetch(url, {
        method: 'DELETE',
        body: JSON.stringify(data),
        headers: {
          "Content-Type": "application/json"
        }
      })

      if (!response.ok) {
        throw Error(`Could not remove ${type}`)
      }

      await dispatch(clearThunkHelper(type)())

      type === 'journal' || type === 'milestone'
        ? dispatch(fetchDataThunk(data.user_id, data.project_id, type))
        : dispatch(fetchDataThunk(data.user_id, '', type))

    } catch (error) {
      dispatch(setError(error.message))
      return false
    }
  }
}