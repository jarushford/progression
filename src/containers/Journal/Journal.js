import React from 'react'
import { deleteDataThunk } from '../../thunks/deleteData'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

export function Journal({ journal_date, entry, id, user_id, project_id, deleteJournal }) {
  return (
    <article className="journal">
      <div>
        <h3>{journal_date}</h3>
        <i className="fas fa-times" onClick={() => deleteJournal(id, user_id, project_id, 'journal')} />
      </div>
      <p className="milestone-cap">{entry}</p>
    </article>
  )
}

export const mapDispatchToProps = (dispatch) => ({
  deleteJournal: (item_id, user_id, project_id, type) => dispatch(deleteDataThunk(item_id, user_id, project_id, type))
})

export default connect(null, mapDispatchToProps)(Journal)

Journal.propTypes = {
  deleteJournal: PropTypes.func.isRequired
}