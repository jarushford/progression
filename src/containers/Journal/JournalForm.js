import React, { Component } from 'react'
import { connect } from 'react-redux'
import { addDataThunk } from '../../thunks/addData'
import { Redirect } from 'react-router-dom'

class JournalForm extends Component {
  constructor() {
    super()
    this.state = {
      entry: '',
      entryAdded: false
    }
  }

  handleChange = (e) => {
    const { name, value } = e.target
    this.setState({ [name]: value })
  }

  handleSubmit = async (e) => {
    e.preventDefault()
    const { user, project } = this.props
    const date = new Date()
    const today = 
    `${('0' + (date.getMonth() + 1)).slice(-2)}/`
      + `${('0' + (date.getDate())).slice(-2)}/`
      + `${date.getFullYear()}`
    const entry = {
      entry: this.state.entry,
      journal_date: today,
      user_id: user.id,
      project_id: project
    }
    const result = await this.props.addJournal(entry, 'journal')
    if (result) {
      this.setState({ entryAdded: true })
    }
  }

  render() {
    const { entry, entryAdded } = this.state

    if (entryAdded) { return <Redirect to={`/projects/${this.props.project}`} /> }

    return (
      <form className="journal-form" onSubmit={this.handleSubmit}>
        <h2>ADD JOURNAL ENTRY</h2>
        <textarea name="entry"
          value={entry}
          onChange={this.handleChange}
          placeholder="entry"
          maxLength="800"
          className="journal-entry"
        />
        <button>Add Entry</button>
      </form>
    )
  }
}

const mapStateToProps = (state) => ({
  user: state.currentUser,
  project: state.currentProject
})

const mapDispatchToProps = (dispatch) => ({
  addJournal: (entry, type) => dispatch(addDataThunk(entry, type))
})

export default connect(mapStateToProps, mapDispatchToProps)(JournalForm)

