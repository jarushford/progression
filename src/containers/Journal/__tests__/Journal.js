import React from 'react'
import { shallow } from 'enzyme'
import { deleteDataThunk } from '../../../thunks/deleteData'
import { mapDispatchToProps, Journal } from '../Journal'

jest.mock('../../../thunks/deleteData')

describe('Journal', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<Journal />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should delete an entry on click', () => {
    const mockDelete = jest.fn()
    const wrapper = shallow(<Journal deleteJournal={mockDelete} id={1} project_id={2} user_id={4} />)

    wrapper.find('.fa-times').simulate('click')

    expect(mockDelete).toBeCalledWith(1, 4, 2, 'journal')
  })

  it('should return a props object with a method deleteJournal', () => {
    const mockDispatch = jest.fn()
    const expected = deleteDataThunk(1, 2, 3, 'journal')

    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.deleteJournal(1, 2, 3, 'journal')

    expect(mockDispatch).toBeCalledWith(expected)
  })
})