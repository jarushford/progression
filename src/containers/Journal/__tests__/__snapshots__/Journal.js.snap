// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Journal should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Journal
    deleteJournal={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <h3 />
          <i
            className="fas fa-times"
            onClick={[Function]}
          />
        </div>,
        <p
          className="milestone-cap"
        />,
      ],
      "className": "journal",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3 />,
            <i
              className="fas fa-times"
              onClick={[Function]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fas fa-times",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "milestone-cap",
        },
        "ref": null,
        "rendered": null,
        "type": "p",
      },
    ],
    "type": "article",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <h3 />
            <i
              className="fas fa-times"
              onClick={[Function]}
            />
          </div>,
          <p
            className="milestone-cap"
          />,
        ],
        "className": "journal",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3 />,
              <i
                className="fas fa-times"
                onClick={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fas fa-times",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "milestone-cap",
          },
          "ref": null,
          "rendered": null,
          "type": "p",
        },
      ],
      "type": "article",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
